Support MPP Keysend

alecchendev opened the pull request:
Closes #1222. Planning to do a bit more testing, but otherwise I think most of the changes are here.

alecchendev commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -169,7 +169,12 @@ pub(super) fn build_onion_payloads(path: &Vec<RouteHop>, total_msat: u64, paymen
 						})
-					} else { None },
+					} else if keysend_preimage.is_some() && value_msat < total_msat {
+						Some(msgs::FinalOnionHopData {
+							payment_secret: PaymentSecret([0u8; 32]),
+							total_msat,
```
Not super sure what's the best way to fill payment_secret here, so suggestions welcome :)

alecchendev replied:
Changed this to set payment secret to be the preimage

alecchendev commented:
Sure sounds good!

alecchendev commented:
And yea I initially thought we didn't need to validate the payment secret since the preimage is serving the same role here, but I realize even then people probably should still be sending all parts with at least the same secret + it'll be good to keep things consistent with normal MPPs.

alecchendev commented:
Thanks for the heads up! Planning to get to this today or tomorrow

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4041,17 +4040,6 @@ where
 			}
 			expected_amt_msat = htlc.total_value_received;
 
-			if let OnionPayload::Spontaneous(_) = &htlc.onion_payload {
```
Added a new field in `UserConfig`. Although why would it risk panicking as opposed to just failing the payment?

valentinewallace reviewed (commented):
Still going through an initial pass, but looks good :)

Not exactly related to the PR, but I think we can consolidate a good chunk of the logic in `process_pending_htlc_forwards` by always calling `check_total_value` for keysends now: https://pastebin.com/mTCLQxaS

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3523,7 +3522,7 @@ where
 										}
 										let ref mut htlcs = &mut claimable_payment.htlcs;
-										if htlcs.len() == 1 {
+										if htlcs.len() == 1 && !$is_keysend {
```
ah true!

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3435,10 +3430,10 @@ where
 									},
-									PendingHTLCRouting::ReceiveKeysend { payment_preimage, payment_metadata, incoming_cltv_expiry } => {
+									PendingHTLCRouting::ReceiveKeysend { payment_data, payment_preimage, payment_metadata, incoming_cltv_expiry } => {
 										let onion_fields = RecipientOnionFields { payment_secret: None, payment_metadata };
```
Missed that, thanks! Also added sending a secret + checking for it in the test to catch that

alecchendev replied:
Note to self: still gotta update docs

valentinewallace commented:
This DRY of `process_pending_forwards` might be a nice fit for this PR 🥹 https://github.com/lightningdevkit/rust-lightning/pull/2156#pullrequestreview-1402253016

valentinewallace commented:
@TheBlueMatt I have some questions on this. 
1. Currently we override the sender-provided `RecipientOnionFields::payment_secret`. That's fine, but I think `send_spontaneous_payment` should only take a `payment_metadata` then. 
2. From looking through the lnd codebase, it seems like they will reject payment secret keysends. So it seems like ideally we should only set it if we're sending a multi-part keysend, and leave it blank otherwise. 

Do these requested changes make sense, before I send alec on a wild goose chase?

valentinewallace commented:
Maybe lnd would be open to upstreaming this change? It looks like a few line change. 

Maybe a stopgap could be to support receiving MPP keysends but not sending?

valentinewallace commented:
We'll also want to update `InvoiceFeatures::for_keysend` to set mpp as optional

alecchendev commented:
For how to handle the `RecipientOnionFields::payment_secret` that gets fed into a spontaneous payment method: I added a fixup commit to actually override whatever is inputted here (before it’d only do this if payment secret wasn’t present), such that for single-part keysend we never send a payment secret and for multi-part keysend we always send a secret. With this + making it clear in `RecipientOnionFields` docs that setting its `payment_secret` field does nothing on keysend (still getting around to this), even if it’s kinda weird to have a useless field in this one scenario, the user at least can’t accidentally do something wrong. And this accomodates lnd rejecting single-part keysends with payment secrets without any extra lnd-specific constructors iiuc. Is this a good way to go about it?

Otherwise, is the problem here lnd-specific or that generally not all implementations support receiving MPP keysend + there’s not a good way to signal support for it (since keysend means no invoice to communicate features + other contexts' feature bits supporting MPP and keysend individually don’t necessarily signal support for MPP keysend, and MPP keysend doesn’t have its own feature bit)? If anyone knows off the top of their head about eclair and cln that’d be appreciated otherwise I can take a look soon.

> We'll also want to update `InvoiceFeatures::for_keysend` to set mpp as optional

I haven't looked at the routing code much, but if our payee doesn't support MPP keysend, will there be a chance that we end up using a multi-part route and getting the payment failed when we should've just used a single path route? Or if there's a single path route would that generally be preferred, in which case setting this feature by default would only really help us out if our payee does support it?

Also added a to do list in the description to keep track of some of the things I still gotta get around to :)

valentinewallace commented:
> I added a fixup commit to actually override whatever is inputted here (before it’d only do this if payment secret wasn’t present), such that for single-part keysend we never send a payment secret and for multi-part keysend we always send a secret.

I'm tentatively with this idea, plus documentation that lnd doesn't accept mpp keysends. 

I'm not sure about an lnd-specific method, because I don't think wallets generally have logic to detect which lightning implementation they're sending to, so how usable would this bespoke api be? 

Also, worth noting that I don't think said api would decrease overall failures rates that much, we'd just fail at pathfinding instead of at the destination (which is preferable, ofc).

alecchendev commented:
Yea sorry for the delay, I'll be getting back to this soon, I should be good to push some changes within the next couple of days based on everything we've discussed

alecchendev commented:
Made a bunch of changes, annotating main things here:
- Previously would override user's input and add the preimage as the secret on MPP keysend - got rid of that, sticks to using whatever user provides in `RecipientOnionFields` ([`3cdb6be`](https://github.com/lightningdevkit/rust-lightning/pull/2156/commits/3cdb6be5897090e1335781af7d9e036702096ce5))
- Previously added allowing no payment secret for MPP keysend from when I thought we didn't need it - got rid of that ([`c97378f`](https://github.com/lightningdevkit/rust-lightning/pull/2156/commits/c97378f04109534ae104196633227e5e655dd9a2)) (will delete these first three commits when I squash)
- Added a parameter in `PaymentParameters::for_keysend` and `InvoiceFeatures::for_keysend` to set mpp optional or not ([`c0067ff`](https://github.com/lightningdevkit/rust-lightning/pull/2156/commits/c0067ff8c0268bf64de5dddd09cdc0baf3a03f5d))
- Added docs to `RecipientOnionFields`
- DRYed up with https://github.com/lightningdevkit/rust-lightning/pull/2156#pullrequestreview-1402253016
- Rebased/fixed conflict

alecchendev commented:
As always open to further feedback, my thinking behind my latest changes:

I opted for just adding a parameter on `PaymentParameters/InvoiceFeatures::for_keysend` as it seemed the most intuitive to me. At the current moment while MPP keysend is not widely supported, it seems best to have the user make the choice of whether to tell their router to find multi-path routes for keysends, as opposed to sort of opting them in/out by default by either changing the `InvoiceFeatures::for_keysend` implementation or adding a separate constructor. They could always manually set the feature bits here but this way seemed to make the options a bit clearer.

And I opted for letting the user choose whether to include a secret or not (instead of overriding like I had previously implemented) to give them the flexibility of sending a secret for a single path keysend if they really want to, and just generally having the user be explicit about whether they mean to send a secret on a spontaneous payment. The docs will hopefully help them make whatever choice is best for their application.

valentinewallace commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -414,9 +414,11 @@ pub struct RecipientOnionFields {
-	/// payments.
+	/// Some implementations may reject spontaneous payments with payment secrets, so you may only
+	/// want to provide a secret for a spontaneous payment if you know your recipient will not
+	/// reject it, or if you're sending a spontaneous MPP. Spontaneous MPP is not widely supported
+	/// by all implementations yet, so you will want to be careful about when you provide a
+	/// payment secret for spontaneous payments in order to avoid failures.
```
I think this last sentence is redundant

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4259,9 +4231,9 @@ where
 				// that there's one payment here and move on.
-				if sources.len() != 1 {
+				if !self.default_configuration.accept_mpp_keysend && sources.len() != 1 {
 					log_error!(self.logger, "Somehow ended up with an MPP spontaneous payment - this should not be reachable!");
 					debug_assert!(false);
 					valid_mpp = false;
```
Oh true! Also I realized it's actually reachable just generally if our config rejects mpp keysend since the config only means we'll reject them here while claiming, so yea logging an error doesn't really make sense

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3591,27 +3595,22 @@ where
-													if total_value >= msgs::MAX_VALUE_MSAT { break; }
-												},
-												_ => unreachable!(),
+											if htlc.total_msat != $total_msat {
```
great idea - I ended up just checking against `claimable_payment.purpose` instead of the onion payload of each htlc + was able to consolidate some other checks

also ended up being able to drop `$is_keysend` from being passed in and now get it from `claimable_htlc.onion_payload`

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3670,41 +3669,10 @@ where
-														fail_htlc!(claimable_htlc, payment_hash);
-													}
-												}
+												check_total_value!(payment_data.as_ref().map(|d| d.payment_secret), Some(preimage), payment_data.as_ref().map_or(claimable_htlc.value, |d| d.total_msat), true);
```
Oops my bad - removed `$total_msat` from being passed into the macro in favor of `claimable_htlc.total_msat` 👍

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7058,6 +7027,7 @@ impl_writeable_tlv_based_enum!(PendingHTLCRouting,
 	},
 	(2, ReceiveKeysend) => {
 		(0, payment_preimage, required),
+		(1, payment_data, option),
```
In the case that this happens, would a user just not be able to downgrade, or would they just need to handle the payment before downgrading?

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4263,11 +4235,8 @@ where
-				if sources.len() != 1 {
-					log_error!(self.logger, "Somehow ended up with an MPP spontaneous payment - this should not be reachable!");
-					debug_assert!(false);
+				if !self.default_configuration.accept_mpp_keysend && sources.len() != 1 {
```
I originally added it in response to https://github.com/lightningdevkit/rust-lightning/pull/2156#discussion_r1160135118 but yea handling on the receive side makes a lot more sense 😅

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2369,12 +2358,19 @@ where
 					}
-
 					PendingHTLCRouting::ReceiveKeysend {
+						payment_data,
```
Ah ok that makes sense, will change this soon

alecchendev replied:
Changes have been made in https://github.com/lightningdevkit/rust-lightning/pull/2156/commits/df4122ba7cd7e01817e4dbae17e5e4951ef4402e

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3551,12 +3548,20 @@ where
 											events::PaymentPurpose::InvoicePayment {
 												payment_preimage: $payment_preimage,
-												payment_secret: $payment_data.payment_secret,
+												payment_secret: $payment_secret.expect("Should never call check_total_value with non-keysend payment but no payment secret"),
```
Mm good point, realized I could just pass in `purpose` to `check_total_value` and avoid these unwraps altogether, will push these changes soon

alecchendev replied:
Okay, made this change, as well as made `is_keysend` based on this `purpose` that is passed into `check_total_value`, and both of those match the `claimable_htlc.onion_payload` based on where they're called

Also fixed an additional ambiguity ([here](https://github.com/lightningdevkit/rust-lightning/pull/2156/commits/250ee32779d19f7c2426ffd89e102610ffbcfbd1#diff-645d12c7269d09caab57d2e0c1fd34f672ee259994c86ad4bad7f4da9183671cL3691-R3689)) - for the legacy path where we'd store inbound payments, it looked at `payment_data.is_none()` to check if the payment was keysend; added a check on the onion payload since with mpp keysend `payment_data` could be `Some`, and in this path we should only be handling invoice payments

alecchendev commented:
CI is gonna fail until I squash since I made some changes to the tests (eventually fixed in https://github.com/lightningdevkit/rust-lightning/pull/2156/commits/0b35ab877488d43a246e4606501ff5afc2c4ef9e)

Btw thanks a ton for all the review as always @TheBlueMatt @valentinewallace!

alecchendev commented on lightning/src/util/config.rs:
```diff
@@ -552,6 +552,17 @@ pub struct UserConfig {
+	///
+	/// Setting this to true will break backwards compatibility upon downgrading to an LDK
+	/// version < 0.0.116 while receiving an MPP keysend. If we have already received an MPP
+	/// keysend, downgrading will cause [`ChannelManager::claim_funds`] to reject the payment.
```
Oops, thanks.

alecchendev commented:
Squashed! Also rewrote some of the commit messages/descriptions to make them a bit more intuitive, but the overall commits remained the same

alecchendev commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8657,6 +8633,53 @@ mod tests {
+		let updates = get_htlc_update_msgs!(nodes[1], nodes[0].node.get_our_node_id());
+		assert!(updates.update_add_htlcs.is_empty());
+		assert!(updates.update_fulfill_htlcs.is_empty());
+		assert_eq!(updates.update_fail_htlcs.len(), 1);
```
Like it should only fail when `accept_mpp_keysend = true`? As long as we haven't claimed the payment yet, it should fail either way, right?

alecchendev replied:
I think the test above, in combination with `test_keysend_msg_with_secret_err` actually cover it? If our config says we don't accept mpp keysend, if it doesn't have a payment secret then an mpp keysend htlc is just a duplicate payment hash payment and gets caught in the above test, and if it does have a payment secret we just reject it off the bat.

alecchendev replied:
Just added some comments to better express this

alecchendev replied:
Ahh okay gotcha, good catch. Just added a test `test_reject_mpp_keysend_htlc` for this, reverted my previously added comments, and added a few minor touch ups

valentinewallace commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -414,9 +414,9 @@ pub struct RecipientOnionFields {
-	/// Note that for spontaneous payments most lightning nodes do not currently support MPP
-	/// receives, thus you should generally never be providing a secret here for spontaneous
-	/// payments.
+	/// Some implementations may reject spontaneous payments with payment secrets, so you may only
+	/// want to provide a secret for a spontaneous payment if you know your recipient will not
+	/// reject it, or if you're sending a spontaneous MPP.
```
nit: 
```suggestion
	/// Some implementations may reject spontaneous payments with payment secrets, so you may only
	/// want to provide a secret for a spontaneous payment if MPP is needed and you know your recipient
	/// will not reject it.
```

valentinewallace commented on lightning/src/routing/router.rs:
```diff
@@ -621,8 +621,14 @@ impl PaymentParameters {
+	///
+	/// Note that MPP keysend is not widely supported yet. The `allow_mpp` lets you choose
+	/// whether your router will be allowed to find a multi-part route for this payment. If you
+	/// set `allow_mpp` to true, you should also handle sending a payment secret.
```
```suggestion
	/// set `allow_mpp` to true, you should ensure a payment secret is set on send, likely via [`RecipientOnionFields::secret_only`].
```

valentinewallace reviewed (approved):
Just some docs nits, LGTM! Sorry for the delay

alecchendev commented:
Thanks for all the review!
